File: app/diagnostico.py
"""Motor de diagnostico por sintomas - IMPLEMENTACAO DE REFERENCIA.

Recebe uma cultura e um conjunto de sintomas observados, e devolve as doencas
ordenadas por compatibilidade.

NOTA DE PROJETO
---------------
A pontuacao aqui NAO e uma probabilidade. E um indice de compatibilidade
(variante ponderada do indice de Tversky). Chamar isso de "confianca de 92%"
seria desonesto: o sistema nao tem modelo probabilistico por tras. Quando o
modelo de imagem entrar no app (fase 5), ele vai produzir uma confianca real,
e as duas coisas vao conviver como sinais distintos.

Este arquivo e a referencia. O app roda o porte em TypeScript
(web/lib/diagnostico.ts), que e testado contra as fixtures geradas daqui
(app/fixtures.py) para que as duas implementacoes nao divirjam.

Por isso toda a saida e DETERMINISTICA: nenhuma lista sai de iteracao de set,
e todo empate tem criterio de desempate explicito. Sem isso, a ordem mudaria
entre execucoes do Python e nunca bateria com a do TypeScript.

A funcao `diagnosticar` e o ponto de extensao: na fase 5 ela ganha um parametro
opcional com as probabilidades vindas da CNN, e combina os dois sinais.
"""

import unicodedata
from dataclasses import dataclass, field

from app.db import conectar

# Quanto penalizar um sintoma que o usuario marcou mas que a doenca nao explica.
# 0 = ignora sintomas sobrando; 1 = pesa tanto quanto um sintoma faltando.
PENALIDADE_SINTOMA_NAO_EXPLICADO = 0.5

# Abaixo disso a hipotese e fraca demais para ser mostrada.
LIMIAR_MINIMO = 0.15

ROTULOS_GRAVIDADE = {
    1: "Muito baixa", 2: "Baixa", 3: "Media", 4: "Alta", 5: "Muito alta",
}


@dataclass(frozen=True)
class SintomaPontuado:
    """Sintoma dentro do perfil de uma doenca, com o peso que ele tem la."""
    id: str
    nome: str
    peso: float


@dataclass(frozen=True)
class SintomaRef:
    """Sintoma marcado que a doenca em questao nao explica - nao tem peso."""
    id: str
    nome: str


@dataclass
class Hipotese:
    """Uma doenca candidata, com o porque da pontuacao."""
    doenca_id: str
    nome: str
    agente: str
    tipo_agente: str
    gravidade: int
    compatibilidade: float
    classe_modelo: str | None = None
    sintomas_compativeis: list[SintomaPontuado] = field(default_factory=list)
    sintomas_esperados_ausentes: list[SintomaPontuado] = field(default_factory=list)
    sintomas_nao_explicados: list[SintomaRef] = field(default_factory=list)

    @property
    def compatibilidade_pct(self) -> int:
        return round(self.compatibilidade * 100)

    @property
    def rotulo_gravidade(self) -> str:
        return ROTULOS_GRAVIDADE[self.gravidade]


@dataclass(frozen=True)
class Pergunta:
    """A proxima observacao a fazer no campo.

    `descarta` so vem preenchido quando o sintoma e de fato discriminante:
    esperado pela primeira hipotese e ausente do perfil da segunda. Nesse caso
    a resposta separa as duas de uma vez. Quando as duas esperam o mesmo
    sintoma, a pergunta ainda ajuda a confirmar a lider, mas nao desempata -
    e a interface nao deve prometer que desempata.
    """
    sintoma_id: str
    nome: str
    confirma: str
    descarta: str | None = None


def chave_alfabetica(texto: str) -> str:
    """Chave de ordenacao alfabetica para portugues.

    Ordenar por codigo de caractere joga todo acento para depois do 'z':
    "Pessego" viria depois de "Pimentao" e "Acaros" iria para o fim da lista.
    Tirar os diacriticos antes de comparar devolve a ordem que o usuario
    espera. O porte em TypeScript faz exatamente isto (NFD + descarte dos
    combinantes), senao as duas listas sairiam em ordens diferentes.
    """
    sem_acento = "".join(
        c for c in unicodedata.normalize("NFD", texto)
        if not unicodedata.combining(c)
    )
    return sem_acento.casefold()


def listar_culturas(apenas_com_doencas: bool = False) -> list[dict]:
    """As culturas da base.

    `apenas_com_doencas` filtra as que o modelo conhece so como saudaveis
    (mirtilo e framboesa no PlantVillage): oferece-las no fluxo por sintomas
    seria um beco sem saida.
    """
    con = conectar()
    # Colunas explicitas, e nao `c.*`: `prefixo_modelo` e contrato do modelo
    # de imagem, nao dado do fluxo por sintomas, e nao tem por que atravessar
    # ate o seletor de cultura.
    sql = """SELECT c.id, c.nome, c.nome_cientifico, c.emoji,
                    COUNT(d.id) AS n_doencas
               FROM cultura c LEFT JOIN doenca d ON d.cultura_id = c.id
              GROUP BY c.id"""
    if apenas_com_doencas:
        sql += " HAVING n_doencas > 0"
    linhas = con.execute(sql).fetchall()
    con.close()
    return sorted((dict(l) for l in linhas),
                  key=lambda c: chave_alfabetica(c["nome"]))


def listar_sintomas_da_cultura(cultura_id: str) -> list[dict]:
    """So os sintomas que aparecem em alguma doenca daquela cultura.

    Evita mostrar ao usuario sintomas irrelevantes para o que ele plantou -
    das 47 entradas do catalogo, cada cultura usa entre 2 e 20.
    """
    con = conectar()
    linhas = con.execute(
        """SELECT DISTINCT s.id, s.nome, s.orgao,
                  o.rotulo AS orgao_rotulo, o.ordem AS orgao_ordem
             FROM sintoma s
             JOIN orgao o ON o.id = s.orgao
             JOIN doenca_sintoma ds ON ds.sintoma_id = s.id
             JOIN doenca d ON d.id = ds.doenca_id
            WHERE d.cultura_id = ?""",
        (cultura_id,),
    ).fetchall()
    con.close()
    return sorted((dict(l) for l in linhas),
                  key=lambda s: (s["orgao_ordem"], chave_alfabetica(s["nome"])))


def diagnosticar(cultura_id: str, sintomas_marcados: set[str]) -> list[Hipotese]:
    """Pontua todas as doencas da cultura contra os sintomas observados.

                              acertos
    compatibilidade = ---------------------------------
                      acertos + faltantes + ruido * 0.5
    """
    if not sintomas_marcados:
        return []

    con = conectar()
    doencas = con.execute(
        "SELECT * FROM doenca WHERE cultura_id = ? ORDER BY id", (cultura_id,)
    ).fetchall()

    perfis = con.execute(
        """SELECT ds.doenca_id, ds.sintoma_id, ds.peso
             FROM doenca_sintoma ds
             JOIN doenca d ON d.id = ds.doenca_id
            WHERE d.cultura_id = ?""",
        (cultura_id,),
    ).fetchall()

    nomes_sintomas = {
        l["id"]: l["nome"]
        for l in con.execute("SELECT id, nome FROM sintoma").fetchall()
    }
    con.close()

    # Um id que nao existe no catalogo e descartado, nao contado como ruido.
    # O caso real e o app servido de um cache antigo pelo service worker,
    # marcando um sintoma que a base ja removeu: trata-lo como ruido
    # penalizaria todas as hipoteses por igual e estragaria o diagnostico
    # inteiro por causa de um id obsoleto.
    sintomas_marcados = {s for s in sintomas_marcados if s in nomes_sintomas}
    if not sintomas_marcados:
        return []

    # doenca_id -> {sintoma_id: peso}
    perfil_por_doenca: dict[str, dict[str, float]] = {}
    for l in perfis:
        perfil_por_doenca.setdefault(l["doenca_id"], {})[l["sintoma_id"]] = l["peso"]

    hipoteses = []
    for d in doencas:
        perfil = perfil_por_doenca.get(d["id"], {})
        if not perfil:
            continue

        # sorted() em tudo: a ordem de um set em Python depende do hash das
        # strings e mudaria entre execucoes. A soma de floats nao e associativa,
        # entao ordem instavel daria compatibilidades diferentes a cada rodada
        # - e nunca identicas as do TypeScript.
        compativeis = _ordenar_por_peso(
            sorted(sintomas_marcados & perfil.keys()), perfil, nomes_sintomas)
        ausentes = _ordenar_por_peso(
            sorted(perfil.keys() - sintomas_marcados), perfil, nomes_sintomas)
        nao_explicados = [
            SintomaRef(id=s, nome=nomes_sintomas[s])
            for s in sorted(sintomas_marcados - perfil.keys())
        ]

        acertos = sum(s.peso for s in compativeis)
        faltantes = sum(s.peso for s in ausentes)
        ruido = len(nao_explicados) * PENALIDADE_SINTOMA_NAO_EXPLICADO

        denominador = acertos + faltantes + ruido
        score = acertos / denominador if denominador else 0.0

        if score < LIMIAR_MINIMO:
            continue

        hipoteses.append(Hipotese(
            doenca_id=d["id"],
            nome=d["nome"],
            agente=d["agente"],
            tipo_agente=d["tipo_agente"],
            gravidade=d["gravidade"],
            compatibilidade=score,
            classe_modelo=d["classe_modelo"],
            sintomas_compativeis=compativeis,
            sintomas_esperados_ausentes=ausentes,
            sintomas_nao_explicados=nao_explicados,
        ))

    # Empate em compatibilidade: sobe a mais grave, porque o custo de ignorar
    # a doenca mais destrutiva e maior. Empate tambem na gravidade: o id, que
    # nao significa nada mas garante ordem identica nas duas implementacoes.
    hipoteses.sort(key=lambda h: (-h.compatibilidade, -h.gravidade, h.doenca_id))
    return hipoteses


def _ordenar_por_peso(
    ids: list[str], perfil: dict[str, float], nomes: dict[str, str]
) -> list[SintomaPontuado]:
    """Maior peso primeiro; o id desempata para a ordem ser reproduzivel."""
    return [
        SintomaPontuado(id=s, nome=nomes[s], peso=perfil[s])
        for s in sorted(ids, key=lambda x: (-perfil[x], x))
    ]


def melhor_pergunta(hipoteses: list[Hipotese]) -> Pergunta | None:
    """O sintoma que o agronomo deve ir verificar na planta agora.

    Entre os sintomas que a hipotese lider espera e que ainda nao foram
    marcados, escolhe o de maior `peso_na_lider - peso_na_segunda`.

    Essa diferenca e exatamente o quanto a resposta afasta as duas hipoteses:
    se o sintoma estiver presente, ambas ganham, mas a lider ganha a mais
    justamente essa diferenca. Um sintoma que as duas esperam com o mesmo peso
    tem diferenca zero - confirma o quadro e nao decide nada.

    Nao basta pegar o de maior peso: o sintoma mais caracteristico da lider
    costuma ser tambem caracteristico da concorrente, que e por que as duas
    estao empatadas. Nao basta so exigir que a segunda nao o espere: um
    sintoma ocasional exclusivo separa menos que um sintoma classico que a
    concorrente espera fraco.

    Sem segunda colocada, a diferenca vira o proprio peso e a regra se reduz
    a "pergunte o sintoma mais caracteristico".
    """
    if not hipoteses:
        return None

    lider = hipoteses[0]
    if not lider.sintomas_esperados_ausentes:
        return None

    segunda = hipoteses[1] if len(hipoteses) > 1 else None
    perfil_segunda: dict[str, float] = {}
    if segunda is not None:
        for s in segunda.sintomas_compativeis + segunda.sintomas_esperados_ausentes:
            perfil_segunda[s.id] = s.peso

    escolhido = min(
        lider.sintomas_esperados_ausentes,
        # min() com sinais invertidos, e o id por ultimo, para dar a mesma
        # escolha em Python e em TypeScript quando ha empate.
        key=lambda s: (-(s.peso - perfil_segunda.get(s.id, 0.0)), -s.peso, s.id),
    )

    # So promete descartar a segunda quando ela nao espera esse sintoma de
    # forma alguma. Se ela tambem o espera, ainda que mais fraco, dizer que a
    # resposta a descarta seria exagero - e a interface repete o que o motor
    # afirma aqui.
    descarta = (
        segunda.nome
        if segunda is not None and escolhido.id not in perfil_segunda
        else None
    )
    return Pergunta(
        sintoma_id=escolhido.id,
        nome=escolhido.nome,
        confirma=lider.nome,
        descarta=descarta,
    )


def detalhar_doenca(doenca_id: str) -> dict:
    """Ficha completa: descricao, condicoes, tratamentos e ingredientes ativos."""
    con = conectar()
    d = con.execute(
        """SELECT d.*, c.nome AS cultura_nome, c.emoji
             FROM doenca d JOIN cultura c ON c.id = d.cultura_id
            WHERE d.id = ?""",
        (doenca_id,),
    ).fetchone()

    if d is None:
        con.close()
        raise KeyError(f"doenca desconhecida: {doenca_id}")

    tratamentos = con.execute(
        """SELECT tipo, descricao FROM tratamento
            WHERE doenca_id = ?
            ORDER BY CASE tipo
                       WHEN 'cultural'  THEN 1
                       WHEN 'biologico' THEN 2
                       WHEN 'quimico'   THEN 3
                       ELSE 4 END, id""",
        (doenca_id,),
    ).fetchall()

    ingredientes = con.execute(
        "SELECT nome, grupo, acao FROM ingrediente_ativo WHERE doenca_id = ?"
        " ORDER BY id",
        (doenca_id,),
    ).fetchall()
    con.close()

    return {
        "id": d["id"],
        "nome": d["nome"],
        "cultura": d["cultura_nome"],
        "emoji": d["emoji"],
        "classe_modelo": d["classe_modelo"],
        "agente": d["agente"],
        "tipo_agente": d["tipo_agente"],
        "gravidade": d["gravidade"],
        "descricao": d["descricao"],
        "condicoes_favoraveis": {
            "temperatura": d["cond_temperatura"],
            "umidade": d["cond_umidade"],
            "observacao": d["cond_observacao"],
        },
        "tratamentos": [dict(t) for t in tratamentos],
        "ingredientes_ativos": [dict(i) for i in ingredientes],
    }

File: web/lib/diagnostico.ts
/**
 * Motor de diagnostico por sintomas - porte do motor de referencia em Python
 * (app/diagnostico.py).
 *
 * Roda inteiro no navegador, sobre a base embutida no bundle: nenhuma rede,
 * nenhum servidor. E o que faz o fluxo por sintomas funcionar no meio do
 * talhao, em modo aviao.
 *
 * A pontuacao NAO e probabilidade. E um indice de compatibilidade (variante
 * ponderada do indice de Tversky) entre o quadro observado e o perfil tipico
 * da doenca. Quando o modelo de imagem entrar (fase 5), ele produzira uma
 * confianca de verdade, e os dois sinais vao conviver rotulados de forma
 * distinta.
 *
 * As duas implementacoes tem que produzir exatamente os mesmos numeros. O
 * teste em diagnostico.test.ts confere isso contra as fixtures geradas pelo
 * Python. Tres armadilhas de portabilidade estao tratadas e comentadas ao
 * longo do arquivo: ordem de soma de floats, arredondamento de .5 e ordem
 * alfabetica com acento.
 */

import {
  CULTURAS,
  ORGAOS,
  SINTOMAS,
  type Cultura,
  type Doenca,
  type Gravidade,
  type OrgaoId,
} from "./base-conhecimento.ts";

/** Quanto penalizar um sintoma marcado que a doenca nao explica.
 *  0 = ignora sintomas sobrando; 1 = pesa tanto quanto um sintoma faltando. */
export const PENALIDADE_SINTOMA_NAO_EXPLICADO = 0.5;

/** Abaixo disso a hipotese e fraca demais para ser mostrada. */
export const LIMIAR_MINIMO = 0.15;

export const ROTULOS_GRAVIDADE: Record<Gravidade, string> = {
  1: "Muito baixa",
  2: "Baixa",
  3: "Media",
  4: "Alta",
  5: "Muito alta",
};

export type SintomaPontuado = { id: string; nome: string; peso: number };
export type SintomaRef = { id: string; nome: string };

export type Hipotese = {
  doencaId: string;
  nome: string;
  agente: string;
  tipoAgente: string;
  gravidade: Gravidade;
  compatibilidade: number;
  classeModelo: string | null;
  /** Marcados que a doenca explica, do mais caracteristico ao menos. */
  sintomasCompativeis: SintomaPontuado[];
  /** Tipicos da doenca que nao foram marcados - as perguntas em aberto. */
  sintomasEsperadosAusentes: SintomaPontuado[];
  /** Marcados que esta doenca nao explica - o que enfraquece a hipotese. */
  sintomasNaoExplicados: SintomaRef[];
  compatibilidadePct: number;
  rotuloGravidade: string;
};

export type Pergunta = {
  sintomaId: string;
  nome: string;
  confirma: string;
  /** So vem preenchido quando a resposta de fato separa as duas primeiras. */
  descarta: string | null;
};

export type CulturaResumida = {
  id: string;
  nome: string;
  nomeCientifico: string;
  emoji: string;
  nDoencas: number;
};

export type SintomaDoCatalogo = {
  id: string;
  nome: string;
  orgao: OrgaoId;
  orgaoRotulo: string;
  orgaoOrdem: number;
};

export type Ficha = {
  id: string;
  nome: string;
  cultura: string;
  emoji: string;
  classeModelo: string | null;
  agente: string;
  tipoAgente: string;
  gravidade: Gravidade;
  descricao: string;
  condicoesFavoraveis: { temperatura: string; umidade: string; observacao: string };
  tratamentos: { tipo: string; descricao: string }[];
  ingredientesAtivos: { nome: string; grupo: string; acao: string }[];
};

/* ---------------------------------------------------------------- indices */

const NOME_DO_SINTOMA = new Map(SINTOMAS.map((s) => [s.id, s.nome]));
const ORGAO_POR_ID = new Map(ORGAOS.map((o) => [o.id, o]));
const CULTURA_POR_ID = new Map(CULTURAS.map((c) => [c.id, c]));

const DOENCA_POR_ID = new Map<string, { doenca: Doenca; cultura: Cultura }>(
  CULTURAS.flatMap((cultura) =>
    cultura.doencas.map((doenca) => [doenca.id, { doenca, cultura }] as const),
  ),
);

/** doencaId -> { sintomaId: peso } */
const PERFIL_POR_DOENCA = new Map<string, Map<string, number>>(
  [...DOENCA_POR_ID].map(([id, { doenca }]) => [
    id,
    new Map(doenca.sintomas.map((s) => [s.id, s.peso])),
  ]),
);

/* ----------------------------------------------------------------- ordem  */

/**
 * Chave de ordenacao alfabetica para portugues.
 *
 * Comparar strings direto joga todo acento para depois do "z": "Pessego"
 * viria depois de "Pimentao" e "Acaros" iria para o fim da lista. Tirar os
 * diacriticos antes de comparar devolve a ordem que o usuario espera.
 *
 * Nao usar localeCompare aqui: o resultado dele depende do locale do
 * aparelho, e a mesma lista sairia em ordens diferentes em celulares
 * diferentes - e nunca igual a do Python.
 */
export function chaveAlfabetica(texto: string): string {
  return texto
    .normalize("NFD")
    .replace(/\p{Mn}/gu, "")
    .toLowerCase();
}

/** Comparacao de strings por codigo de caractere, igual a do Python. */
function compararTexto(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Soma de floats compensada (Neumaier), que e o que o `sum()` do CPython faz.
 *
 * Somar 0.7 + 0.6 + 0.5 + 0.3 ingenuamente da 2.0999999999999996; o `sum()`
 * do Python devolve 2.1. A diferenca e de um bit e nao muda porcentagem
 * nenhuma na tela, mas quebra a igualdade exata com as fixtures - e uma
 * comparacao por tolerancia deixaria passar justamente as divergencias reais
 * que este teste existe para pegar.
 *
 * O Python e a implementacao de referencia, entao quem se ajusta e o porte.
 * A compensacao guarda o erro de arredondamento de cada soma parcial e o
 * devolve no fim.
 */
function somar(valores: number[]): number {
  let total = 0.0;
  let compensacao = 0.0;
  for (const valor of valores) {
    const parcial = total + valor;
    compensacao +=
      Math.abs(total) >= Math.abs(valor)
        ? total - parcial + valor
        : valor - parcial + total;
    total = parcial;
  }
  return total + compensacao;
}

/**
 * Arredondamento do Python: meio para o par ("banker's rounding").
 *
 * Math.round(12.5) da 13 e round(12.5) do Python da 12. Uma compatibilidade
 * que caia exatamente em .5 mostraria porcentagens diferentes nas duas
 * implementacoes - e o teste de fixtures acusaria uma divergencia real.
 */
function arredondarMeioParaPar(valor: number): number {
  const piso = Math.floor(valor);
  const resto = valor - piso;
  if (resto > 0.5) return piso + 1;
  if (resto < 0.5) return piso;
  return piso % 2 === 0 ? piso : piso + 1;
}

/* ---------------------------------------------------------------- catalogo */

/**
 * As culturas da base.
 *
 * `apenasComDoencas` filtra as que o PlantVillage so conhece como saudaveis
 * (mirtilo e framboesa): oferece-las no fluxo por sintomas seria um beco sem
 * saida.
 */
export function listarCulturas(apenasComDoencas = false): CulturaResumida[] {
  return CULTURAS.filter((c) => !apenasComDoencas || c.doencas.length > 0)
    .map((c) => ({
      id: c.id,
      nome: c.nome,
      nomeCientifico: c.nomeCientifico,
      emoji: c.emoji,
      nDoencas: c.doencas.length,
    }))
    .sort((a, b) =>
      compararTexto(chaveAlfabetica(a.nome), chaveAlfabetica(b.nome)),
    );
}

/**
 * So os sintomas que aparecem em alguma doenca daquela cultura.
 *
 * Das 58 entradas do catalogo, cada cultura usa entre 4 e 26. Mostrar as 58
 * transformaria a tela num formulario impossivel de ler sob sol.
 */
export function listarSintomasDaCultura(culturaId: string): SintomaDoCatalogo[] {
  const cultura = CULTURA_POR_ID.get(culturaId);
  if (!cultura) return [];

  const usados = new Set(
    cultura.doencas.flatMap((d) => d.sintomas.map((s) => s.id)),
  );

  return SINTOMAS.filter((s) => usados.has(s.id))
    .map((s) => {
      const orgao = ORGAO_POR_ID.get(s.orgao)!;
      return {
        id: s.id,
        nome: s.nome,
        orgao: s.orgao,
        orgaoRotulo: orgao.rotulo,
        orgaoOrdem: orgao.ordem,
      };
    })
    .sort(
      (a, b) =>
        a.orgaoOrdem - b.orgaoOrdem ||
        compararTexto(chaveAlfabetica(a.nome), chaveAlfabetica(b.nome)),
    );
}

/* -------------------------------------------------------------- pontuacao */

/**
 * Pontua todas as doencas da cultura contra os sintomas observados.
 *
 *                           acertos
 * compatibilidade = ---------------------------------
 *                   acertos + faltantes + ruido * 0.5
 *
 * O denominador penaliza os dois erros possiveis: quadro incompleto
 * (faltantes) e quadro contaminado (ruido).
 */
export function diagnosticar(
  culturaId: string,
  sintomasMarcados: Iterable<string>,
): Hipotese[] {
  const cultura = CULTURA_POR_ID.get(culturaId);
  if (!cultura) return [];

  // Um id fora do catalogo e descartado, nao contado como ruido. O caso real
  // e o service worker servindo um bundle antigo que marca um sintoma ja
  // removido da base: trata-lo como ruido penalizaria todas as hipoteses por
  // igual e estragaria o diagnostico inteiro por causa de um id obsoleto.
  const marcados = new Set(
    [...sintomasMarcados].filter((s) => NOME_DO_SINTOMA.has(s)),
  );
  if (marcados.size === 0) return [];

  const hipoteses: Hipotese[] = [];

  for (const doenca of cultura.doencas) {
    const perfil = PERFIL_POR_DOENCA.get(doenca.id)!;

    const compativeis = ordenarPorPeso(
      [...perfil.keys()].filter((s) => marcados.has(s)),
      perfil,
    );
    const ausentes = ordenarPorPeso(
      [...perfil.keys()].filter((s) => !marcados.has(s)),
      perfil,
    );
    const naoExplicados = [...marcados]
      .filter((s) => !perfil.has(s))
      .sort(compararTexto)
      .map((s) => ({ id: s, nome: NOME_DO_SINTOMA.get(s)! }));

    // Somar na MESMA ordem e pelo MESMO algoritmo do Python. Soma de ponto
    // flutuante nao e associativa: 0.4+0.3+0.9 e 0.9+0.4+0.3 diferem no
    // ultimo bit. As duas listas ja vieram ordenadas por (-peso, id), que e
    // a ordem usada la, e `somar` replica a compensacao do `sum()` do Python.
    const acertos = somar(compativeis.map((s) => s.peso));
    const faltantes = somar(ausentes.map((s) => s.peso));
    const ruido = naoExplicados.length * PENALIDADE_SINTOMA_NAO_EXPLICADO;

    const denominador = acertos + faltantes + ruido;
    const compatibilidade = denominador ? acertos / denominador : 0;

    if (compatibilidade < LIMIAR_MINIMO) continue;

    hipoteses.push({
      doencaId: doenca.id,
      nome: doenca.nome,
      agente: doenca.agente,
      tipoAgente: doenca.tipoAgente,
      gravidade: doenca.gravidade,
      compatibilidade,
      classeModelo: doenca.classeModelo,
      sintomasCompativeis: compativeis,
      sintomasEsperadosAusentes: ausentes,
      sintomasNaoExplicados: naoExplicados,
      compatibilidadePct: arredondarMeioParaPar(compatibilidade * 100),
      rotuloGravidade: ROTULOS_GRAVIDADE[doenca.gravidade],
    });
  }

  // Empate em compatibilidade: sobe a mais grave, porque o custo de ignorar
  // a doenca mais destrutiva e maior. Empate tambem na gravidade: o id, que
  // nao significa nada mas garante ordem identica nas duas implementacoes.
  return hipoteses.sort(
    (a, b) =>
      b.compatibilidade - a.compatibilidade ||
      b.gravidade - a.gravidade ||
      compararTexto(a.doencaId, b.doencaId),
  );
}

/** Maior peso primeiro; o id desempata para a ordem ser reproduzivel. */
function ordenarPorPeso(
  ids: string[],
  perfil: Map<string, number>,
): SintomaPontuado[] {
  return ids
    .map((id) => ({ id, nome: NOME_DO_SINTOMA.get(id)!, peso: perfil.get(id)! }))
    .sort((a, b) => b.peso - a.peso || compararTexto(a.id, b.id));
}

/**
 * O sintoma que o agronomo deve ir verificar na planta agora.
 *
 * Entre os sintomas que a hipotese lider espera e que ainda nao foram
 * marcados, escolhe o de maior `peso na lider - peso na segunda`.
 *
 * Essa diferenca e exatamente o quanto a resposta afasta as duas hipoteses:
 * se o sintoma estiver presente, ambas ganham, mas a lider ganha a mais
 * justamente essa diferenca. Um sintoma que as duas esperam com o mesmo peso
 * tem diferenca zero - confirma o quadro e nao decide nada.
 *
 * Nao basta pegar o de maior peso: o sintoma mais caracteristico da lider
 * costuma ser tambem caracteristico da concorrente, que e por que as duas
 * estao empatadas.
 */
export function melhorPergunta(hipoteses: Hipotese[]): Pergunta | null {
  if (hipoteses.length === 0) return null;

  const lider = hipoteses[0];
  if (lider.sintomasEsperadosAusentes.length === 0) return null;

  const segunda = hipoteses.length > 1 ? hipoteses[1] : null;
  const perfilSegunda = new Map<string, number>(
    segunda
      ? [...segunda.sintomasCompativeis, ...segunda.sintomasEsperadosAusentes].map(
          (s) => [s.id, s.peso],
        )
      : [],
  );

  const ganho = (s: SintomaPontuado) => s.peso - (perfilSegunda.get(s.id) ?? 0);

  const escolhido = [...lider.sintomasEsperadosAusentes].sort(
    (a, b) => ganho(b) - ganho(a) || b.peso - a.peso || compararTexto(a.id, b.id),
  )[0];

  // So promete descartar a segunda quando ela nao espera esse sintoma de
  // forma alguma. Se ela tambem o espera, ainda que mais fraco, dizer que a
  // resposta a descarta seria exagero - e a interface repete o que o motor
  // afirma aqui.
  const descarta =
    segunda && !perfilSegunda.has(escolhido.id) ? segunda.nome : null;

  return {
    sintomaId: escolhido.id,
    nome: escolhido.nome,
    confirma: lider.nome,
    descarta,
  };
}

/** Ficha completa: descricao, condicoes, manejo e ingredientes ativos. */
export function detalharDoenca(doencaId: string): Ficha {
  const encontrado = DOENCA_POR_ID.get(doencaId);
  if (!encontrado) throw new Error(`doenca desconhecida: ${doencaId}`);

  const { doenca, cultura } = encontrado;

  // Cultural antes de biologico antes de quimico. Nao e cosmetico: e a ordem
  // que o manejo integrado recomenda, e inverte-la sugeriria que a primeira
  // resposta a uma doenca e pulverizar.
  const ordemMip: Record<string, number> = { cultural: 1, biologico: 2, quimico: 3 };
  const tratamentos = [...doenca.tratamentos].sort(
    (a, b) => (ordemMip[a.tipo] ?? 4) - (ordemMip[b.tipo] ?? 4),
  );

  return {
    id: doenca.id,
    nome: doenca.nome,
    cultura: cultura.nome,
    emoji: cultura.emoji,
    classeModelo: doenca.classeModelo,
    agente: doenca.agente,
    tipoAgente: doenca.tipoAgente,
    gravidade: doenca.gravidade,
    descricao: doenca.descricao,
    condicoesFavoraveis: doenca.condicoesFavoraveis,
    tratamentos: tratamentos.map((t) => ({ tipo: t.tipo, descricao: t.descricao })),
    ingredientesAtivos: doenca.ingredientesAtivos.map((i) => ({
      nome: i.nome,
      grupo: i.grupo,
      acao: i.acao,
    })),
  };
}
